File: scala/src/main/scala/operationabstraction/scalaexample/Basic.scala
package operationabstraction.scalaexample

object Basic:
  trait ManufacturingCost:
    def materialCost(quantity: Int): Int
    def assemblyCost(quantity: Int): Int

  def calculateTotalCost(manufacturingCost: ManufacturingCost, quantity: Int): Int =
    val material = manufacturingCost.materialCost(quantity)
    val assembly = manufacturingCost.assemblyCost(quantity)
    material + assembly

  object StandardManufacturingCost extends ManufacturingCost:
    def materialCost(quantity: Int): Int =
      quantity * 100

    def assemblyCost(quantity: Int): Int =
      quantity * 50

  def result: Int =
    calculateTotalCost(StandardManufacturingCost, 10)

File: fsharp/Basic.fs
module Basic

let calculateTotalCost
    (materialCost: int -> int)
    (assemblyCost: int -> int)
    (quantity: int)
    : int =
    let material = materialCost quantity
    let assembly = assemblyCost quantity
    material + assembly

let standardMaterialCost: int -> int =
    fun quantity -> quantity * 100

let standardAssemblyCost: int -> int =
    fun quantity -> quantity * 50

let result =
    calculateTotalCost standardMaterialCost standardAssemblyCost 10
